src/Equations/SPIHandler.cpp

```diff
@@ -315,50 +583,112 @@ void SPIHandler::Rebuild(real_t dt){
+                    if (YpPrevious[ip]>0 && irp[ip]<nr){
+                        YpConversion(ip);
+                        AssignShardSpecificDriftParameters(ip);
+                        AssignDriftComputationParameters(ip);
+                        AssignDriftTimeParameters(ip);
+                        shift_r[ip] = Deltar(ip);
+                        nbrShiftGridCell[ip] = CalculateDriftIrp(ip, shift_r[ip]);// Negative if shift is towards smaller radii
+                        shift_store[ip] = shift_r[ip];
+                    }
                 }
-            }else if(rCoordPNext[ip]<rCoordPPrevious[ip]){
-                for(len_t ir=nr-1;ir>nbrShiftGridCell[ip];ir--){
-                    depositionProfilesAllShards[ir*nShard+ip]=rGrid->GetVpVol(ir-nbrShiftGridCell[ip])/rGrid->GetVpVol(ir)*depositionProfilesAllShards[(ir-nbrShiftGridCell[ip])*nShard+ip];
+            }
+        } else if(spi_shift_mode==OptionConstants::EQTERM_SPI_SHIFT_MODE_PRESCRIBED || spi_deposition_mode==OptionConstants::EQTERM_SPI_DEPOSITION_MODE_LOCAL_LAST_FLUX_TUBE){// Prescribed drift (in terms of grid cells)
+            // If the radial projection of the drift is negative, the shift should go towards smaller radii, 
+            // unless the shift goes past the core and ends at a larger radii on the other side. Here we 
+            // account for this when setting the prescribed shift, keeping in mind that nbrShiftGridCell 
+            // should be negative if the shift is towards smaller radii, while nbrShiftGridCellPrescribed 
+            // is always positive
+            for(len_t ip=0;ip<nShard;ip++){
+                if(cosThetaDrift[ip]<0){// Is the radial projection of the drift negative?
+                    nbrShiftGridCell[ip] = std::abs((int_t)irp[ip]-nbrShiftGridCellPrescribed[ip])-(int_t)irp[ip];
+                    
+                    //Account for that grid cell 0 should be counted twice (on both sides of the magnetic axis)
+                    if((int_t)irp[ip]-nbrShiftGridCellPrescribed[ip]<0)
+                        nbrShiftGridCell[ip]--;    
+                }else if(rCoordPNext[ip]<rCoordPPrevious[ip]){
+                    nbrShiftGridCell[ip] = nbrShiftGridCellPrescribed[ip];
                 }
             }
         }
-
-    }else if(spi_deposition_mode==OptionConstants::EQTERM_SPI_DEPOSITION_MODE_LOCAL_LAST_FLUX_TUBE){
-        CalculateTimeAveragedDeltaSourceLocal(depositionProfilesAllShards);
-
-        // Shift the deposition profile to the last grid cell (before the current one) to avoid "self-dilution"
+        // Shift the deposition profile
         for(len_t ip=0;ip<nShard;ip++){
-            if(rCoordPNext[ip]>rCoordPPrevious[ip]){
-                for(len_t ir=0;ir<nr-1;ir++){
-                    depositionProfilesAllShards[ir*nShard+ip]=rGrid->GetVpVol(ir+1)/rGrid->GetVpVol(ir)*depositionProfilesAllShards[(ir+1)*nShard+ip];
+            if(nbrShiftGridCell[ip]<0){
```

Comment:
Somehow, the code is getting to this point without ``nbrShiftGridCell`` being initialized (it is allocated, but its elements have not been set to any particular values).